File: lincoln/ruby/lib/lang/plan/if_then_else.rb
class IfThenElse < Expression
	def IfThenElse(type, ifexpr, thenexpr, elseexpr)
		@type = type
		@ifexpr = ifexpr
		@thenexpr = thenexpr
		@elseexpr = elseexpr
	end

	def to_s
		return ifexpr.to_s + " ? " +  thenexpr.to_s + " : " + elseexpr.to_s
	end

  attr_reader :type

  def variables
		variables = Array.new
		ifexpr.variables.each { |v| variables << v }
		thenexpr.variables.each { |v| variables << v }
		elseexpr.variables.each { |b| variables << v }
		return variables
	end

  def function()
    test = @ifexpr.function
    thencase = thenexpr.function
    elsecase = elseexpr.function

    e_lam = lambda do |t|
      return (text.evaluate(t) == true ? thencase.evaluate(t) : elsecase.evaluate(t))
    end

    r_lam = lambda do
      return @type
    end

    tmpClass = Class.new(TupleFunction)
    tmpClass.send :define_method, :evaluate do |tuple|
      e_lam.call(tuple)
    end
    tmpClass.send :define_method, :returnType do 
      r_lam.call
    end
    return tmpClass.new
  end
end

File: lincoln/java/src/jol/lang/plan/IfThenElse.java
package jol.lang.plan;

import java.util.HashSet;
import java.util.Set;

import jol.types.basic.Tuple;
import jol.types.exception.P2RuntimeException;
import jol.types.function.TupleFunction;

public class IfThenElse extends Expression {
	
	private Class type;
	
	private Boolean ifexpr;
	
	private Expression thenexpr;
	
	private Expression elseexpr;

	public IfThenElse(Class type, Boolean ifexpr, Expression thenexpr, Expression elseexpr) {
		this.type = type;
		this.ifexpr = ifexpr;
		this.thenexpr = thenexpr;
		this.elseexpr = elseexpr;
	}

	@Override
	public String toString() {
		return ifexpr.toString() + " ? " + 
		       thenexpr.toString() + " : " + 
		       elseexpr.toString();
	}

	@Override
	public Class type() {
		return this.type;
	}

	@Override
	public Set<Variable> variables() {
		Set<Variable> variables = new HashSet<Variable>();
		variables.addAll(ifexpr.variables());
		variables.addAll(thenexpr.variables());
		variables.addAll(elseexpr.variables());
		return variables;
	}

	@Override
	public TupleFunction function() {
		return new TupleFunction() {
			private final TupleFunction<java.lang.Boolean> test = ifexpr.function();
			private final TupleFunction thencase = thenexpr.function();
			private final TupleFunction elsecase = elseexpr.function();
			public Object evaluate(Tuple tuple) throws P2RuntimeException {
				return test.evaluate(tuple).equals(java.lang.Boolean.TRUE) ? 
						thencase.evaluate(tuple) : elsecase.evaluate(tuple);
			}

			public Class returnType() {
				return type();
			}
		};
	}

}
